### DISCUSSION TITLE
[general] crew might fail when number of targets is large (roughly larger than 1000)

### DISCUSSION THREAD

**Discussion by psychelzh** [ropensci/targets#1217] [General]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I am sorry I cannot provide a simole reprex for this. But for a large project, e.g., [my pipeline script](https://github.com/psychelzh/cogstruct-dev/blob/main/_scripts/predict_phenotypes.R), if {crew} is used for parallel analysis, I constantly get the following error:

```
! Error running targets::tar_make()
Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
Debugging guide: https://books.ropensci.org/targets/debugging.html
How to ask for help: https://books.ropensci.org/targets/help.html
Last error message:
    {crew} worker 15 launched 5 times in a row without completing any tasks. Either troubleshoot or raise launch_max asLast error traceback:
```

But if I use `targets::tar_make_future()`, it never fails. I guess it is because {crew} dispatches all the targets though pending.

**Comment by wlandau** [comment]
Unfortunately we can't say what exactly is happening. If you can provide a simpler runnable reprex, or if you can isolate the problem to `crew` without using `targets`, that would help a lot.

**Reply by wlandau** [reply]
I guess you might try installing development `crew` (c.f. https://github.com/wlandau/crew/issues/79#issuecomment-1906115892) and also ruling out out-of-memory issues in your computing environment.

**Reply by psychelzh** [reply]
Thanks. I tried the development version, but in vain. Currently, only `tar_make_future()` works perfectly, and the other two, {crew} and {clustermq}, both failed at some point. But I am very sorry again that providing a reprex seems rather hard for me.

**Reply by wlandau** [reply]
Some ideas:

1. `clustermq` had a rough couple of versions around 0.9.0, but version 0.9.3 should work in most situations. Might be worth a shot.
2. `clustermq` and `crew` use network programming, whereas `future.batchtools` does not. You might try the `mirai`-specific test at https://github.com/wlandau/crew.cluster/discussions/36#discussioncomment-8224120 to ensure parallel workers can connect to your local R session over the local network of your cluster.
3. If the test in (2) passes, then the `"{crew} worker 15 launched 5 times in a row` error might be caused by tasks that uses a large amounts of memory and crashes the workers. Out-of-memory errors are hard to detect, and the problem can be exacerbated if a worker runs multiple tasks. In`crew`, you could try `tasks_max = 1` to make sure each task gets a fresh new worker. IIRC `clustermq` has no such functionality yet.
4. A test I often do on my company's SGE cluster is:

```r
# _targets.R file
library(crew.cluster)
library(targets)
controller <- crew_controller_sge(
  workers = 25,
  tasks_max = 100,
  script_lines = "module load R/4.2.2"
)
tar_option_set(controller = controller)
list(
  tar_target(x, seq_len(10000)),
  tar_target(y, Sys.sleep(1), pattern = map(x))
)
```

If this reproduces the same error on your SGE cluster, the next step could be to see how simple you can make it and still reproduce the error. You could try removing `tasks_max`, reducing the number of workers, reducing the number of dynamic branches, setting `Sys.sleep()` to a shorter sleep duration, etc. 
5. You could try testing `crew` without `targets`:

```r
controller <- crew.cluster::crew_controller_sge(
  workers = 25,
  tasks_max = 100,
  script_lines = "module load R/4.2.2"
)
controller$map(
  command = Sys.sleep(1),
  iterate = list(x = seq_len(10000))
)
```

If it reproduces the error, then it could be something to do with your cluster, your installation of NNG, etc. If it works fine with no errors, then it could be something to do with `targets`.

**Reply by psychelzh** [reply]
Thank you very much for your explanation! For my circumstances, it turns out to be the (3) case. After I add `tasks_max = 1`, it works without error.

**Reply by wlandau** [reply]
Fantastic! That all makes sense, and I am glad it works for you now.

**Reply by wlandau** [reply]
For what it's worth, `reset_globals = TRUE` might be an alternative to `tasks_max = 1` for reducing memory costs in some cases, although the latter is still perfectly reasonable and should be safer in some circumstances.

**Reply by philiporlando** [reply]
Would the same `"{crew} worker 15 launched 5 times in a row"` error apply when using `crew::crew_controller_local()`?

**Reply by wlandau** [reply]
It might, depending on what else is going on in your pipeline. At any rate, it is straightforward to try `reset_globals = TRUE` and see if it works for you.

**Comment by psychelzh** [comment]
Recently, I found that if I have roughly 10k targets in a project, and {crew} scheduler will consume a crazy 40 GiB memory. But that is not an issue for the superseded `tar_make_future()` function. I know there are many advantages using {crew}, for example, better schedule planning, but given superseding means it will cover what the superseded function can do, I am wondering if the similar behavior like `tar_make_future()` will be possible?

Although it is probably not good practice for having so many targets in a project, the fact is that data analysis goes into many branches very quickly! Therefore, the behavior like `tar_make_future()` could sometimes save our life a lot. Or can there be more efficient memory compressing for {crew}.

**Reply by psychelzh** [reply]
Or following the logic that superseded functions will never go away, I should just switch to `tar_make_future()` for such circumstances?

**Reply by wlandau** [reply]
Either should be fine. The memory issue could be due to https://github.com/shikokuchuo/mirai/discussions/55#discussioncomment-9057611, in which case updating to development `crew` may fix it.

**Reply by wlandau** [reply]
`storage = "worker"`, `retrieval = "worker"`, `garbage_collection = TRUE`, and `memory = "transient"` should also reduce memory usage.
